### ISSUE TITLE
Improve supported `$PSDefaultParameterValues` command types

### ISSUE THREAD

**Issue by surfingoldelephant** [PowerShell/PowerShell#24708]
### Summary of the new feature / enhancement

This issue supersedes https://github.com/PowerShell/PowerShell/issues/24697, which was focused on the lack of support for aliases that target the _path_ of a `.ps1` script file.

The purpose of this issue is to improve [`$PSDefaultParameterValues`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_parameters_default_values) command type support _in general_, which goes beyond the scope of the original issue.

Currently, `$PSDefaultParameterValues` supports/has the following limitations:
* Cmdlets and advanced functions/filters:
  * **Cannot** be module-qualified.
  * **Cannot** include `:` in the name.
* Advanced `.ps1` script files:
  * Must be specified by name (with extension), as reported by `(Get-Command -Name .\script.ps1).Name`.
  * **Cannot** be a full or relative path.
* Aliases: 
  * The alias _name_:
    * Can be module-qualified (unlike with cmdlet/function names).
    * **Cannot** include `:` in the name.
  * The alias _target_ (as reported by `AliasInfo.Definition`):
    * Can be a cmdlet or advanced function/filter name.
    * Can be an advanced script file _name_. 
    * Can include `:` in the name.
    * **Cannot** be another alias name.
    * **Cannot** be module-qualified.
    * **Cannot** be a full or relative script file path.
* In all cases, the command/target command:
  * **Cannot** be non-advanced/simple (must be a cmdlet or have `CmdletBindingAttribute`/`ParameterAttribute` decoration). 
  * Can be `Import-Module -Prefix`-generated.
  * Can include uncommon characters such as spaces in the name (except `:` in the cases above).

As the above highlights, there are numerous limitations/inconsistencies, some of which are undocumented as of writing. See:
* https://github.com/MicrosoftDocs/PowerShell-Docs/issues/11604

To remove inconsistencies and align with PowerShell's typical command calling semantics, `$PSDefaultParameterValues` should support the following:

* Command names with `:`.
* `.ps1` script files specified as a full or relative path (including as the target of an alias).
* Module-qualified command names.
* Non-advanced/simple commands (this may not be deemed worthwhile considering the technical differences between cmdlet and script parameter binding).

### Proposed technical implementation details (optional)

Given the following example commands:

```powershell
'param ([Parameter()] $Foo) $Foo' | 
    Set-Content -LiteralPath Temp:\ScriptFile.ps1

function Func:tion { param ([Parameter()] $Foo) $Foo }

function SimpleFunction($Foo) { $Foo }

function AdvancedFunction { param ([Parameter()] $Foo) $Foo }

Set-Alias -Name ScriptFilePathAlias -Value Temp:\ScriptFile.ps1

Set-Alias -Name ScriptFileNameAlias -Value ScriptFile.ps1
$Env:PATH += "{0}{1}" -f [IO.Path]::PathSeparator, (Get-Item -LiteralPath Temp:).Fullname
```

As of v7.5.0, no default value is supplied by the following `$PSDefaultParameterValues` usage. Support should be added for each of the scenarios outlined.

```powershell
# Advanced script file by full path.
$PSDefaultParameterValues = @{ 'Temp:\ScriptFile.ps1:Foo' = 'Bar' }
Temp:\ScriptFile.ps1

# Advanced script file by relative path.
$PSDefaultParameterValues = @{ '.\ScriptFile.ps1:Foo' = 'Bar' }
Set-Location -LiteralPath Temp:
.\ScriptFile.ps1

# Alias whose target is an advanced script file by full/relative path.
$PSDefaultParameterValues = @{ 'ScriptFilePathAlias:Foo' = 'Bar' }
ScriptFilePathAlias

# Cmdlet or advanced command with ":" in name.
$PSDefaultParameterValues = @{ 'Func:tion:Foo' = 'Bar' }
Func:tion

# Module-qualified cmdlet or advanced function/filter.
$PSDefaultParameterValues = @{ 'Microsoft.PowerShell.Management\Get-Item:Path' = 'Bar' }
Microsoft.PowerShell.Management\Get-Item
Get-Item

# Non-advanced/simple command.
# May be out-of-scope for this proposal.
$PSDefaultParameterValues = @{ 'SimpleFunction:Foo' = 'Bar' }
SimpleFunction
```

For reference, a default value _is supplied_ in the following examples:

```powershell
# Advanced script by name only.
$PSDefaultParameterValues = @{ 'ScriptFile.ps1:Foo' = 'Bar' }
ScriptFile.ps1

# Alias whose target is an advanced command by name.
$PSDefaultParameterValues = @{ 'ScriptFile.ps1:Foo' = 'Bar' }
ScriptFileNameAlias

# Module-unqualified cmdlet or advanced function/filter
$PSDefaultParameterValues = @{ 'Get-Item:Path' = 'Bar' }
Microsoft.PowerShell.Management\Get-Item
Get-Item

$PSDefaultParameterValues = @{ 'AdvancedFunction:Foo' = 'Bar' }
AdvancedFunction
```

**Comment by MatejKafka** [comment]
> Limits the usefulness considerably as script files currently must be $Env:PATH-discoverable.

Not exactly – the name is matched even if you invoke the script using an absolute/relative path:
```powershell
$PSDefaultParameterValues["asdf.ps1:Arg"] = "Value"
D:\test\asdf.ps1 # this receives -Arg Value
```

Nevertheless, I agree that adding support for absolute paths to `.ps1` scripts would be pretty useful.

**Comment by surfingoldelephant** [comment]
> the name is matched even if you invoke the script using an absolute/relative path:

Yes, you're absolutely correct. In fact, I have that covered at the bottom of the currently supported scenarios.

```powershell
$PSDefaultParameterValues = @{ 'ScriptFile.ps1:Foo' = 'Bar' }
Temp:\ScriptFile.ps1 # Bar
```

The note regarding ```$Env:PATH``` discoverability was misplaced. It was intended to be under the alias section (now corrected - thank you!).

To clarify, adding an alias that targets a script file to `$PSDefaultParameterValues` is only supported if the target (`AliasInfo.Definition`) is a file _name_. This isn't particularly useful given calling the alias will only succeed if the target script file is ```$Env:PATH```-discoverable.
